### ISSUE TITLE
Late subscribers to Multicast don't complete

### ISSUE THREAD

**Issue by bholloway** [ReactiveX/RxJS#1217]
**Scenario**

* Multicast a **cold** Observable using `publish()` and `refCount()` based on a source Observable.
* Source Observable notifies COMPLETE.
* New subscriber `subscribe()`s to the multicast observable.

**Expected Behavior**

* Late subscriber immediately receives a COMPLETE notification

**Exhibited Behavior**

* Late subscriber receives no notifications

**Rationale**

* The COMPLETE notification is important to indicate an Observable is in a disposed state, where the [Observable contract](http://reactivex.io/documentation/contract.html) is discontinued.
* The COMPLETE **state** is communicated as an **event**. Any eventful system with non-persistent elements will loose state.
* Even if the system described above is composed early and retained it is not persistent, because its connectivity is not established until it becomes **hot**. Therefore there is no communication of the disposed **state** downstream to the late subscriber.

**Example**

This `.lifecycle()` operator [specification](https://github.com/bholloway/rx-operators#operatorlifecyclescheduler--observable) and [implementation](https://github.com/bholloway/rx-operators/blob/master/operator/lifecycle.js).

Test failures

[rx-operators/operator/lifecycle.spec.js:101:41](https://github.com/bholloway/rx-operators/blob/master/operator/lifecycle.spec.js#L101)
```
  1) lifecycle lifecycle observable notify COMPLETE should occur when the upstream observable is already COMPLETE
```

[rx-operators/operator/lifecycle.spec.js:180:41](https://github.com/bholloway/rx-operators/blob/master/operator/lifecycle.spec.js#L180)
```
  2) lifecycle downstream observable notify COMPLETE should occur when the upstream observable is already COMPLETE
```

**Comment by trxcllnt** [comment]
@bholloway can you post steps to reproduce? Running the following example in [requirebin](http://requirebin.com/) executes correctly:

```es6
var Rx = require('rxjs');
var source = Rx.Observable
	.range(0, 10)
	.finally(console.log.bind(console, "unsubscribe"))
	.publish().refCount();

source.subscribe(null, null, console.log.bind(console, "1. done"));
source.subscribe(null, null, console.log.bind(console, "2. done"));

/* prints:
> 1. done
> unsubscribe
> 2. done
*/
```

**Comment by bholloway** [comment]
Thanks @trxcllnt for your quick response.

I am probably flawed in the technique that I use to trigger `next()` and `complete()` using `stimulus` variable. However here is my example.

http://requirebin.com/?gist=828dd9d724ebb0a1caf4

```javascript
var Rx = require('rxjs');

// -------------------- A : complete then subscribe : FAIL --------------------------------
// get an observable similar to how we would create a data source
var stimulusA;
var sourceA = Rx.Observable.create(function (instance) {
    stimulusA = instance;
  })
	.publish()
	.refCount();

// hack to ensure that stimulus is assigned for the purposes of this example
sourceA.subscribe().unsubscribe();
console.log('stimulus-A', typeof stimulusA);

// complete then subscribe 
stimulusA.complete();
sourceA.subscribe(undefined, undefined, console.log.bind(console, 'complete-A')); 

/*
stimulus-A object
*/

// ------------------- B : subscribe then complete : PASS --------------------------------
// get an observable similar to how we would create a data source
var stimulusB;
var sourceB = Rx.Observable.create(function (instance) {
    stimulusB = instance;
  })
	.publish()
	.refCount();

// hack to ensure that stimulus is assigned for the purposes of this example
sourceB.subscribe().unsubscribe();
console.log('stimulus-B', typeof stimulusB);

// subscribe then complete
sourceB.subscribe(undefined, undefined, console.log.bind(console, 'complete-B'));
stimulusB.complete();

/*
stimulus-B object
complete-B
*/
```

**Comment by bholloway** [comment]
@trxcllnt I think this differs from your example because `range()` is more like `defer()` than `create()`.

You will notice in [`range.ts:57`](https://github.com/ReactiveX/RxJS/blob/master/src/observable/range.ts#L57) that the range knows that it is complete and sends `complete()` to the late subscriber.

Here is a similar implementation with `defer()` and internal state `isDisposed`.

http://requirebin.com/?gist=45b2d35782eb4df2b628

```javascript
var Rx = require('rxjs');

// -------------------- A : complete then subscribe : PASS --------------------------------
// get an observable similar to how we would create a data source
var stimulusA;
var isDisposedA = false;
var commonA = Rx.Observable.create(function (instance) {
    stimulusA = instance;
  })
var sourceA = Rx.Observable.defer(function () {
    return isDisposedA ? Rx.Observable.empty() : commonA;  
  })
	.publish()
	.refCount();

// hack to ensure that stimulus is assigned
sourceA.subscribe().unsubscribe();
console.log('stimulus-A', typeof stimulusA);

// complete then subscribe
isDisposedA = true;
stimulusA.complete(); 
sourceA.subscribe(undefined, undefined, console.log.bind(console, 'complete-A')); 

/*
stimulus-A object
complete-A
*/

// ------------------- B : subscribe the complete : PASS --------------------------------
// get an observable similar to how we would create a data source
var stimulusB;
var isDisposedB = false;
var commonB = Rx.Observable.create(function (instance) {
    stimulusB = instance;
  })
  var sourceB = Rx.Observable.defer(function () {
    return isDisposedB ? Rx.Observable.empty() : commonB;  
  })
	.publish()
	.refCount();

// hack to ensure that stimulus is assigned
sourceB.subscribe().unsubscribe();
console.log('stimulus-B', typeof stimulusB);

// subscribe then complete
sourceB.subscribe(undefined, undefined, console.log.bind(console, 'complete-B'));
isDisposedB = true;
stimulusB.complete();

/*
stimulus-B object
complete-B
*/
```

I am suggesting that this is fundamental to the [Observable contract](http://reactivex.io/documentation/contract.html).

Perhaps `create()` is very low level and does not implement this. However, as a subscriber, I am dependent on how well the source Observables are created in whichever 3rd party library I am using.

**Comment by trxcllnt** [comment]
@bholloway ah, I see what you're trying to do. Rx is working as expected, so let me try to break down what you should do differently (and why) without sounding like a condescending jerk:

Observables are lazy by default, so each time they're subscribed to, their subscribe function (the callback you pass to `Observable.create`) is called with a new Observer instance. `publish` (and all other variants of `multicast`) short-circuit this behavior by inserting a Subject between the source Observable and `n` Observers. The source Observable thinks there's only one Subscriber (the Subject), and it's the Subject's job to send the source's events to multiple places:
```
                    |--> observer 1
source --> subject -|--> observer 2
                    |--> ... observer n
```

`refCount` stores the number of subscribers, and subscribes/unsubscribes the shared Subject to/from the source when the number of subscribers grows beyond/shrinks to zero. When the line I've marked as `subscription 1` executes, `refCount` is indeed subscribing to your source Observable (writing the Observer to closure state). You're probably thinking that Observer is the shared Subject instance, but it isn't (this is a departure from the RxJS v4 behavior). Since RxJS v5 has merged the Observer and Subscription types into a single type (Subscriber), we have to wrap Subjects inside a Subscriber when they're used as Observers.

So when you save out the Observer inside your create function, you're actually saving out a Subscriber that wraps the Subject. By immediately calling `unsubscribe`, you're disposing of that `Subscriber`, which means the Subscriber ignores your subsequent call to `complete()`, and *doesn't* forward on the `complete` to the Subject.

When `subscription 2` executes, `refCount` resubscribes the internal `Subject` to the source (which also means that `stimulusA` gets written out again), but the internal `Subject` hasn't heard anything.

```es6
// -------------------- A : complete then subscribe : FAIL --------------------------------
// get an observable similar to how we would create a data source
var stimulusA;
var sourceA = Rx.Observable.create(function (instance) {
    stimulusA = instance;
  })
    .publish()
    .refCount();

// hack to ensure that stimulus is assigned for the purposes of this example
/*subscription 1 */ sourceA.subscribe().unsubscribe(); /* <-- this sets `isUnsubscribed` on the internal Subject */
console.log('stimulus-A', typeof stimulusA);

// complete then subscribe 
stimulusA.complete();
/*subscription 2 */ sourceA.subscribe(undefined, undefined, console.log.bind(console, 'complete-A')); 
```

So I'd recommend two changes:

1. Don't share Observers passed to your Observable's `subscribe` calls. Observers are memory-isolated, and only safe to reference inside the scope of the `subscribe` function.
2. If you need external control over your multicast Observables, create the Subject externally and pass it in via `multicast`.

```es6
var stimulusA = new Rx.Subject();
var sourceA = someSourceObservable
  .multicast(stimulusA)
  .refCount();

stimulusA.complete();
sourceA.subscribe(undefined, undefined, console.log.bind(console, 'complete-A'));
// 'complete-A'
```

[Here's](http://requirebin.com/?gist=ae2154771f58b541c54c) another example in requirebin. It looks like the (order of unsubscription)[https://github.com/ReactiveX/RxJS/blob/master/src/observable/ConnectableObservable.ts#L91] is wrong (should be `1. done`, `unsubscribe`, `2. done`), so I'll file a bug against `refCount`.

**Comment by bholloway** [comment]
Wow that last example is a cool way to get explicit control. Much better than the hack I was doing. I need to finish my unit tests and refactor! Thanks @trxcllnt.

I now see that `create()` is being called for each subscription. It seems obvious now why the argument is named `observer`. However the factory nature of `defer()` led me to expected `create()` to be called **once**.

I now see that the Observserables differentiate subscribers. However I don't yet see how `defer()` is useful - I'm sure it will solve a problem I haven't encountered yet.

**To summarise:**

* Connectivity is driven by subscribe calls on the Observable.
* Some Observables will aggregate and otherwise manage upstream subscription to get a certain result (i.e. multicast).
* Each operator in the chain adds another layer to `_subscribe()`, essentially intercepting/decorating the subscription process from end-to-end.
* The implementation of the Observable needs to ensure that it maintains the contract for each observer that subscribes.

**Comment by trxcllnt** [comment]
@bholloway sounds like you're on the right track. My favorite analogy is [comparing Observables to functions](https://github.com/ReactiveX/RxJS/issues/752#issuecomment-161234549), but it took me at least a year after I started working with Rx to recognize that.

> However the factory nature of defer() led me to expected create() to be called once.

When in doubt, feel free to inspect the [source](https://github.com/ReactiveX/RxJS/blob/master/src/observable/defer.ts#L16). `Observable.defer` is just like `Observable.create`, except the signature of their callback is different:

[Observable.create](https://github.com/ReactiveX/RxJS/blob/master/src/Observable.ts#L75):
```
create<T>(subscribeFunc: (observer: Observer<T>) => Subscription | Function | void) => Observable<T>
```
[`Observable.defer`](https://github.com/ReactiveX/RxJS/blob/master/src/observable/defer.ts#L8):
```
defer<T>(observableFactory: () => Observable<T>) => Observable<T>;
```

`Observable.create`'s `subscribeFunc` accepts an Observer and returns a Subscriber, whereas `Observable.defer`'s function accepts nothing, and returns the Observable to subscribe to. `defer` is used to implement things like `Observable.if`:

```es6
class Observable<T> {
  static if(condition: () => boolean,
            ifSource: Observable<T>,
            elseSource:Observable<R>): Observable<T | R> {
    return Observable.defer(() => {
      return condition() ? ifSource : elseSource;
    });
  }
}
```

Cheers!

**Comment by bholloway** [comment]
@trxcllnt With your guidance I have been able to reimplement my operator in a fraction of the size and it is passing all unit tests. My sincere thanks.
